### PR TITLE
app/mediarecorder: Add BufferOutputDataSource app.

### PR DESCRIPTION
**Pull request by YiHyunJin**
Save data to file by buffer and play the file.

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [96414e4c0c900b222d735e04e6d54fe245c5bd1b] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [96414e4c0c900b222d735e04e6d54fe245c5bd1b] - Code Rule Check (C++) OK.

**Comment by JeongJunSik** [comment]
i think you need to fd close logic in onRecordStopError also.

File: apps/examples/mediarecorder/mediarecorder_main.cpp [hunk 1]
```diff
@@ -91,6 +101,7 @@ class MediaRecorderTest : public MediaRecorderObserverInterface,
 
 	void onRecordBufferDataReached(MediaRecorder& mediaRecorder, unsigned char *data, size_t size)
 	{
+		fwrite(data, sizeof(unsigned char), size, mfp);
```
[the remark this answers is not in the document]
Comment (YiHyunJin) [reply apps/examples/mediarecorder/mediarecorder_main.cpp]: okay i will fix.

**Comment by seinfra** [comment]
Target : [bdb3e2e9c2ac5861fb973a5a2fc415a81378f0e5] - Code Rule Check (C++) OK.

File: apps/examples/mediarecorder/mediarecorder_main.cpp [hunk 2]
```diff
@@ -156,7 +170,7 @@ class MediaRecorderTest : public MediaRecorderObserverInterface,
 				} else if (test == TEST_OPUS) {
 					filePath = "/tmp/record.opus";
 				} else if (test == TEST_BUFFER) {
-					filePath = "";
+					filePath = "/tmp/record.pcm";
```
Comment (Taejun-Kwon) [inline apps/examples/mediarecorder/mediarecorder_main.cpp]: Let's make combination.
```
#define TEST_MEDIATYPE_PCM 
....

#define TEST_DATASOURCE_TYPE_FILE
....

mTest => mMediaType
mDataSourceType => add


if (mDataSourceType == TEST_DATASOURCE_TYPE_FILE) {
    datasource = file;
} else {
    datasource = buffer;
}
....

```

Comment (YiHyunJin) [reply apps/examples/mediarecorder/mediarecorder_main.cpp]: I fixed.

**Comment by seinfra** [comment]
Target : [07825d66a8c092959aaea592b66c07f09077d523] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [07825d66a8c092959aaea592b66c07f09077d523] - Code Rule Check OK.

**Comment by gcjjyy** [comment]
Please resolve conflicts

File: apps/examples/mediarecorder/mediarecorder_main.cpp [hunk 3]
```diff
@@ -151,15 +168,20 @@ class MediaRecorderTest : public MediaRecorderObserverInterface,
...
-				} else if (test == TEST_OPUS) {
+				} else if (mMediaType == TEST_MEDIATYPE_OPUS) {
 					filePath = "/tmp/record.opus";
-				} else if (test == TEST_BUFFER) {
-					filePath = "";
+				} 
+
+				if (mDataSourceType == TEST_DATASOURCE_TYPE_FILE) {
+					mMr.setDataSource(unique_ptr<FileOutputDataSource>(
+						new FileOutputDataSource(2, 16000, AUDIO_FORMAT_TYPE_S16_LE, filePath)));
```
Comment (chanijjani) [inline apps/examples/mediarecorder/mediarecorder_main.cpp]: How about changing "2" and "16000" as macro definitions..?

**Comment by seinfra** [comment]
Target : [14f06dd522197ca2b79c9112c3d14c3f7a1d9404] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [82dd4b358fa7ee1a16466dd6bf6a29736aa54b57] -

**Comment by seinfra** [comment]
Target : [82dd4b358fa7ee1a16466dd6bf6a29736aa54b57] -

**Comment by seinfra** [comment]
Target : [5aed32ba1c24162d8bd17c5f568b3b473fd1187c] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [5aed32ba1c24162d8bd17c5f568b3b473fd1187c] - Code Rule Check (C++) OK.

File: apps/examples/mediarecorder/mediarecorder_main.cpp [hunk 4]
```diff
@@ -150,7 +167,7 @@ class MediaRecorderTest : public MediaRecorderObserverInterface,
 			case APP_ON:
 				std::cout << "SELECTED APP ON" << std::endl;
 				mMr.create();
-				mMr.setObserver(shared_from_this());
+				mMr.setObserver(shared_from_this());
```
[the remark this answers is not in the document]
Comment (YiHyunJin) [reply apps/examples/mediarecorder/mediarecorder_main.cpp]: @junmin-kim  okay!

**Comment by seinfra** [comment]
Target : [9468312c486fd196307630a9ea2adba9ab6a1349] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [9468312c486fd196307630a9ea2adba9ab6a1349] - Code Rule Check (C++) OK.

File: apps/examples/mediarecorder/mediarecorder_main.cpp [hunk 5]
```diff
@@ -150,7 +167,7 @@ class MediaRecorderTest : public MediaRecorderObserverInterface,
...
```
Comment (JeongJunSik) [inline apps/examples/mediarecorder/mediarecorder_main.cpp]: please, remove this empty blank

**Comment by seinfra** [comment]
Target : [1cb1a663478c9588d05ab830db8a09df8dd32e1f] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [1cb1a663478c9588d05ab830db8a09df8dd32e1f] - Code Rule Check OK.

**Pull request merged**